support-frontend/app/utils/CheckoutValidationRules.scala

```diff
@@ -413,8 +423,27 @@ object PaperValidation {
 
   }
 
-  def postcodeIsWithinNationalDeliveryArea(postcode: String): Result = {
-    Valid // TODO: determine valid postcodes
+  def deliveryAgentChosenWhichCoversPostcode(
+      paperRound: PaperRoundAPI,
+      deliveryAgent: Option[Integer],
+      postcode: String,
+  )(implicit
+      ex: ExecutionContext,
+  ): Future[Result] = deliveryAgent match {
+    case None => Future { Invalid(s"User has not chosen a delivery agent") }
+    case Some(agent) =>
+      paperRound
+        .coverage(RequestBody(postcode))
+        .map(response =>
+          response.data.status match {
+            case CO if response.data.agents.map(_.agentId).contains(agent) => Valid
+            case _ =>
+              Invalid(
+                s"User’s postcode $postcode wasn’t covered by their chosen delivery agent $agent"
+                  + s": PaperRound response was $response",
```

Comment:
is this returned to the client or just logged on the server?  We should be a bit careful about including arbitrary content if it might go back to the client

Reply:
Ah ok, that’s a good point – I’ll log the paper round response instead of returning it to the client.